### ISSUE TITLE
[Feature Request] Linux Support

### ISSUE THREAD

**Issue by Henry-Hiles** [mimoguz/tripeaks-gdx#34]
Hello! I'm not sure if it's possible, but it would be great if linux support could be added to this app.

I have a linux phone, and there currently aren't any good solitaire games (and this is the best I've seen for Android). Again, I have no GDX knowledge, so this might be impossible, sorry if so. If you can get it compiling for Linux, I can package it as a Flatpak and put it on Flathub, so it can be downloaded on any linux distro. Thanks!

**Comment by Henry-Hiles** [comment]
Since libGDX is cross-platform, it should be possible, right?

**Comment by mimoguz** [comment]
> Since libGDX is cross-platform, it should be possible, right?

Probably? Apparently LibGDX already supports Linux on Arm, and the project genarator I used, gdx-liftoff, creates basic setup for cross-compiling [lwjgl3](https://github.com/mimoguz/tripeaks-gdx/tree/main/lwjgl3) module using [construo](https://github.com/fourlastor-alexandria/construo). I'd been able to create a Linux x64 runnable by simply running ```./gradlew packageLinuxX64```. I don't know anything about Linux phones though.

**Comment by mimoguz** [comment]
A small upadate: I played with the build config a bit and was able to add Linux AARCH64 as a target. Now ```./gradlew packageLinuxAARCH64``` creates something 🙂. I don't know if that 'something' works though.

**Comment by Henry-Hiles** [comment]
> A small upadate: I played with the build config a bit and was able to add Linux AARCH64 as a target. Now `./gradlew packageLinuxAARCH64` creates something 🙂. I don't know if that 'something' works though.

If you've pushed that build config change I'll try it on my Linux phone later today :)

**Comment by mimoguz** [comment]
> If you've pushed that build config change I'll try it on my Linux phone later today :)

I did.

**Comment by Henry-Hiles** [comment]
I ran it, where does it output to?

**Comment by mimoguz** [comment]
> I ran it, where does it output to?

```lwjgl3/build/construo/<target OS & architecture>/roast/```

**Comment by Henry-Hiles** [comment]
ty

**Comment by Henry-Hiles** [comment]
Hmm.
![Screenshot from 2024-09-02 14-33-55](https://github.com/user-attachments/assets/0e02d05b-8c69-4dfd-bd2e-f37ea7ef4f7b)

**Comment by Henry-Hiles** [comment]
Got a better error: 
![Screenshot from 2024-09-02 14-36-24](https://github.com/user-attachments/assets/8e6bf69f-a40c-40d8-9bd6-86c78707d88a)

**Comment by Henry-Hiles** [comment]
ah
![Screenshot from 2024-09-02 14-41-28](https://github.com/user-attachments/assets/8beb778f-ef1e-4dfd-803d-cb96cc377c3c)

**Comment by mimoguz** [comment]
But it's there, right?

![Screenshot from 2024-09-02 21-45-42](https://github.com/user-attachments/assets/ae73cf17-7ce6-4eda-afa7-9829ff940520)

Or not. It's complaining about ld-linux...

**Comment by Henry-Hiles** [comment]
Looks related to this?: https://github.com/00-Evan/shattered-pixel-dungeon/issues/385#issuecomment-884290949

**Comment by Henry-Hiles** [comment]
What version of libgdx is used

**Comment by Henry-Hiles** [comment]
> But it's there, right?
> 
> ![Screenshot from 2024-09-02 21-45-42](https://private-user-images.githubusercontent.com/7283047/363774450-ae73cf17-7ce6-4eda-afa7-9829ff940520.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzMDM1MzgsIm5iZiI6MTcyNTMwMzIzOCwicGF0aCI6Ii83MjgzMDQ3LzM2Mzc3NDQ1MC1hZTczY2YxNy03Y2U2LTRlZGEtYWZhNy05ODI5ZmY5NDA1MjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDJUMTg1MzU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzVlNGIwMzA4ODFmNzYwNTU5N2RiZTRhODJmMDA1YWU4ZjEzMzIyZTFmZDYyNmViMWNkNjhjMzc3NjFmYWY0OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.U4V3sNuEminpnnSMwsI2YEcC_H-OqRqEKfSbkNy-qF8)
> 
> Or not. It's complaining about ld-linux...

wait a second, thats not there for me

**Comment by Henry-Hiles** [comment]
oh did you unpack the jar

**Comment by mimoguz** [comment]
> oh did you unpack the jar

Yes

**Comment by Henry-Hiles** [comment]
Oh yeah its in the unpacked jar.

**Comment by mimoguz** [comment]
Maybe that? : [https://forums.raspberrypi.com/viewtopic.php?t=357800](https://forums.raspberrypi.com/viewtopic.php?t=357800). 32 bit OS?

**Comment by Henry-Hiles** [comment]
Certainly I'm not on 32 bit, its arm64

**Comment by Henry-Hiles** [comment]
> Looks related to this?: [00-Evan/shattered-pixel-dungeon#385 (comment)](https://github.com/00-Evan/shattered-pixel-dungeon/issues/385#issuecomment-884290949)

Did you see this? What version of libgdx is being used? Also maybe this? https://github.com/libgdx/libgdx/issues/1127

**Comment by Henry-Hiles** [comment]
shattered-pixel-dungeon definitely has a working flatpak linux aarch64 build (ive played it)

**Comment by mimoguz** [comment]
> Did you see this? What version of libgdx is being used? Also maybe this? [libgdx/libgdx#1127](https://github.com/libgdx/libgdx/issues/1127)

```gdxVersion=1.12.1```

Well native libs are there, and linuxX64 build works 😐

**Comment by mimoguz** [comment]
Quite a late and probably unhelpful addition: I finally grabbed a micro-HDMI adapter, hooked my Raspberry Pi 4 to a monitor, and tried the AARCH64 build there. It just worked:
![20240920_21h16m18s_grim](https://github.com/user-attachments/assets/378efba4-1e6b-4a14-bb73-c1314dab9b12)

So, I think the build artifacts are not the issue on your phone.

**Comment by Henry-Hiles** [comment]
Hmm. Ill test again soon, but maybe libgdx requires glibc rather than musl?

**Comment by mimoguz** [comment]
>  maybe libgdx requires glibc rather than musl?

That might be the problem yeah.
[https://github.com/libgdx/libgdx/issues/6983](https://github.com/libgdx/libgdx/issues/6983)

**Comment by Henry-Hiles** [comment]
Great! This means we just need to flatpak it and it'll work I think!

**Comment by Henry-Hiles** [comment]
Is it possible to make a jar for all architechtures?

**Comment by mimoguz** [comment]
> Is it possible to make a jar for all architechtures?

```./gradlew lwjgl3:jar``` should create a multiplatform fat-jar. ```./gradlew lwjgl3:assembleDist``` (or ```distZip```/```distTar```) should create archive files with everything + launch scripts.

**Comment by Henry-Hiles** [comment]
> > Is it possible to make a jar for all architechtures?
> 
> `./gradlew lwjgl3:jar` should create a multiplatform fat-jar. `./gradlew lwjgl3:assembleDist` (or `distZip`/`distTar`) should create archive files with everything + launch scripts.

Sorry, where does that assembleDist output? Thanks!

**Comment by Henry-Hiles** [comment]
Nevermind, found it!

**Comment by Henry-Hiles** [comment]
Hi, I need an application name with at least two periods, whereas I think you usually use ogz.tripeaks? Whats a good 3 part one?

**Comment by Henry-Hiles** [comment]
I'm going to use io.github.mimoguz.tripeaks for now

**Comment by Henry-Hiles** [comment]
Also, it would be ideal if you could make a release that has the lwjgl3-1.2.3.zip as an artifact

**Comment by Henry-Hiles** [comment]
well that was not too bad
![image](https://github.com/user-attachments/assets/498abfe0-1931-4921-bcca-1cf05bae6483)

**Comment by Henry-Hiles** [comment]
Thats amd64 though, ill try it on my phone rn

**Comment by Henry-Hiles** [comment]
Okay: i got it running:
![Screenshot from 2024-12-09 18-52-50](https://github.com/user-attachments/assets/cc1a01fa-66d0-4100-a60b-ca9d9ecf31a9)


However: What is that aspect ratio?? so much green, even if i make it horizontal.

And, more importantly: touch doesnt work.

**Comment by Henry-Hiles** [comment]
Okay, touchscreen works through xwayland.

**Comment by Henry-Hiles** [comment]
Wait, does this app even play sound?

**Comment by Henry-Hiles** [comment]
i think the app normally has no sound which is good. so now, issues are: cant fullscreen app (needs changes in the app, probably a toggle in settings), and huge wasted space (maybe my fault)?
![Screenshot from 2024-12-09 20-00-02](https://github.com/user-attachments/assets/de94852f-83db-460d-b592-ef8079a45393)

**Comment by mimoguz** [comment]
> However: What is that aspect ratio?? so much green, even if i make it horizontal.

Word size is between 304x168 and 360x168, screen size is some integer multiple of that. So strictly landscape. What is the  actual resolution in the latest screenshot?

> Wait, does this app even play sound?

No, no sound :)

**Comment by Henry-Hiles** [comment]
> > However: What is that aspect ratio?? so much green, even if i make it horizontal.
> 
> Word size is between 304x168 and 360x168, screen size is some integer multiple of that. So strictly landscape. What is the actual resolution in the latest screenshot?
> 
> > Wait, does this app even play sound?
> 
> No, no sound :)

the resolution is 2340x1080, but I have 300% fractional scaling.

**Comment by Henry-Hiles** [comment]
Maybe its related to fullscreen, if your app is locked to a certain aspect ratio. Do you think you can add a toggle in settings?

**Comment by mimoguz** [comment]
> Maybe its related to fullscreen, if your app is locked to a certain aspect ratio. Do you think you can add a toggle in settings?

No to the toggle. But I maybe I can add a command line parameter to the desktop launcher.

**Comment by Henry-Hiles** [comment]
> > Maybe its related to fullscreen, if your app is locked to a certain aspect ratio. Do you think you can add a toggle in settings?
> 
> No to the toggle. But I maybe I can add a command line parameter to the desktop launcher.

Why no toggle for fullscreen? I was planning to put this on flathub so people on mobile and desktop can install it, why can't you put a toggle if people aren't on android?

**Comment by Henry-Hiles** [comment]
In my opinion, when linux is detected as the OS, a toggle should be shown in settings and F11 should be bound by the app.

**Comment by mimoguz** [comment]
> In my opinion, when linux is detected as the OS, a toggle should be shown in settings and F11 should be bound by the app.

This would require some platform-dependent code on the core project. And let's face it, desktop linux has much better alternatives to this game already. Anyhow, ```--fullscreen``` command line parameter now starts desktop app in fullscreen mode.

**Comment by Henry-Hiles** [comment]
I'll make it fullscreen by default for the flatpak then

**Comment by Henry-Hiles** [comment]
Okay, next thing would be making an info file for description, name etc. This is a different games one, if you could make one that'd be great: https://github.com/flathub/com.shatteredpixel.shatteredpixeldungeon/blob/master/com.shatteredpixel.shatteredpixeldungeon.metainfo.xml

**Comment by Henry-Hiles** [comment]
> I'll make it fullscreen by default for the flatpak then
fullscreen mostly fixed aspect ratio:
![Screenshot from 2024-12-10 11-53-41](https://github.com/user-attachments/assets/555b59e8-a4d0-4d6e-ab97-4edc1f0af905)

**Comment by Henry-Hiles** [comment]
Here are some guidelines for that file: https://docs.flathub.org/docs/for-app-authors/submission/

I can provide screenshots of the app if you like

**Comment by Henry-Hiles** [comment]
More guidelines: https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/quality-guidelines

**Comment by Henry-Hiles** [comment]
So far as i can see no changes to the app or icon are needed, we can just fill out the metadata.

**Comment by Henry-Hiles** [comment]
Eh, I'll get started on the metainfo myself, I can ask if I need info

**Comment by Henry-Hiles** [comment]
Yeah, I'll just write it, it shouldn't be too bad. I'll do it tonight.

**Comment by Henry-Hiles** [comment]
Okay, metainfo done. Could you please make a new release with the latest changes and with the lwjgl multiplatform zip as an artifact? Thanks.

**Comment by Henry-Hiles** [comment]
Made some metainfo updates, but i still need a fresh release with that zip as an artifact @mimoguz

**Comment by mimoguz** [comment]
> Made some metainfo updates, but i still need a fresh release with that zip as an artifact @mimoguz

Sorry, I was away for a few days. A release? I'll need a tag that F-Droid wouldn't pick up. I'll look into it this weekend.

**Comment by mimoguz** [comment]
> Could you please make a new release with the latest changes and with the lwjgl multiplatform zip as an artifact?

But how do I do that? :)

**Comment by Henry-Hiles** [comment]
> > Could you please make a new release with the latest changes and with the lwjgl multiplatform zip as an artifact?
> 
> But how do I do that? :)

What do you mean? Just make a new release, and upload that zip file as one of the artifacts

**Comment by Henry-Hiles** [comment]
> > Made some metainfo updates, but i still need a fresh release with that zip as an artifact @mimoguz
> 
> Sorry, I was away for a few days. A release? I'll need a tag that F-Droid wouldn't pick up. I'll look into it this weekend.

Oh, does f-droid automatically pick up new releases? I'm not able to find anything about this online

**Comment by Henry-Hiles** [comment]
> > > Made some metainfo updates, but i still need a fresh release with that zip as an artifact @mimoguz
> > 
> > 
> > Sorry, I was away for a few days. A release? I'll need a tag that F-Droid wouldn't pick up. I'll look into it this weekend.
> 
> Oh, does f-droid automatically pick up new releases? I'm not able to find anything about this online

From what I can see, it does this only if the app version has changed in gradle, so you can probably just make a new release but not change any gradle stuff.

**Comment by mimoguz** [comment]
> From what I can see, it does this only if the app version has changed in gradle, so you can probably just make a new release but not change any gradle stuff.

It's configured in the app's metadata file. In my case, I see it's ```Tags ^[0-9.]+$```. So any tag that matches the pattern will be picked up. A tag that contains a letter should be fine, I think.

**Comment by Henry-Hiles** [comment]
> > From what I can see, it does this only if the app version has changed in gradle, so you can probably just make a new release but not change any gradle stuff.
> 
> It's configured in the app's metadata file. In my case, I see it's `Tags ^[0-9.]+$`. So any tag that matches the pattern will be picked up. A tag that contains a letter should be fine, I think.

Hmm, it should really be a proper tag, as this is what will be submitted to flathub. This made me think it only rebuilds if the app version changes in gradle: https://forum.f-droid.org/t/metadata-real-auto-update-when-a-new-tag-is-created/17397

**Comment by mimoguz** [comment]
> What do you mean? Just make a new release, and upload that zip file as one of the artifacts

Oh, that's what you meant. Sorry, my mind went to fat-jars and such. 

Your pull request is still marked as a draft, though. Let me know when it's not.

**Comment by Henry-Hiles** [comment]
> > What do you mean? Just make a new release, and upload that zip file as one of the artifacts
> 
> Oh, that's what you meant. Sorry, my mind went to fat-jars and such.
> 
> Your pull request is still marked as a draft, though. Let me know when it's not.

Oh, whoops, I'll mark it ready, it is ready now. But yeah, it'll need a built lwjgl zip (the zip generated to `lwjgl3/build/distributions/lwjgl3-1.2.3.zip` by `./gradlew lwjgl3:assembleDist`). The PR doesn't need to be merged before the release, the PR is just so we can build flatpaks off of main. The reason I need a release is so that I can put it on flathub, which uses a seperate repo with flathub only stuff.

**Comment by Henry-Hiles** [comment]
All we have to do is this release and then I can start the flathub review process.

**Comment by mimoguz** [comment]
The release is up.

**Comment by Henry-Hiles** [comment]
thanks i will change to use the new icon then make a pr here, and then make a pr to flathub (i didnt see there was a new icon yet)

**Comment by Henry-Hiles** [comment]
oh! you did that, great!

**Comment by Henry-Hiles** [comment]
gonna try to put on flathub then

**Comment by Henry-Hiles** [comment]
Suddenly getting error at the end of build, will look into later.

**Comment by Henry-Hiles** [comment]
Blocked based upon comments here: https://github.com/mimoguz/tripeaks-gdx/commit/26276825c72f45a307934e9821067e2b717e98af

EDIT: Solved, uploading

**Comment by Henry-Hiles** [comment]
I did just have to rename the ID to io.github.mimoguz.Tripeaks-GDX. See https://github.com/flathub/flathub/pull/6023. @mimoguz, maybe you want to do the same here?

**Comment by Henry-Hiles** [comment]
Looks like I'm going to need a few manifest changes. Going to make a PR :)

**Comment by mimoguz** [comment]
I merged the new pr and replaced the release:
[https://github.com/mimoguz/tripeaks-gdx/releases/tag/v1.2.3](https://github.com/mimoguz/tripeaks-gdx/releases/tag/v1.2.3)

**Comment by Henry-Hiles** [comment]
Nice, I need to query something with them and then we're good.

**Comment by mimoguz** [comment]
It seems the game is now published on Flathub. I'm closing the issue.

On a side note: Bloody hell, the Flathub version is 100 times the size of the Android version! 😮

**Comment by Henry-Hiles** [comment]
> It seems the game is now published on Flathub. I'm closing the issue.
> 
> On a side note: Bloody hell, the Flathub version is 100 times the size of the Android version! 😮

Yep, it is!!! Btw, I'm going to open one more issue for you to verify the flatpak (you just need to make an issue after I do something, and then we will get a checkmark)
